pilot/pkg/model/xds_cache.go

```diff
@@ -167,19 +156,21 @@ func NewLenientXdsCache() XdsCache {
 		store:            newLru(),
 		configIndex:      map[ConfigKey]sets.Set{},
 		typesIndex:       map[config.GroupVersionKind]sets.Set{},
-		nextToken:        atomic.NewUint64(0),
 	}
 }
 
 type lruCache struct {
 	enableAssertions bool
 	store            simplelru.LRUCache
-	// nextToken stores the next token to use. The content here doesn't matter, we just need a cheap
-	// unique identifier.
-	nextToken   *atomic.Uint64
-	mu          sync.RWMutex
-	configIndex map[ConfigKey]sets.Set
-	typesIndex  map[config.GroupVersionKind]sets.Set
+	// token stores the latest token of the store, used to prevent stale data overwrite.
+	token CacheToken
+	// It is set when Clear or ClearAll are called, indicates `Add` a cache with same token above is disallowed.
+	// It is reset when after a successful `Add` call, in this case the token will be incremented.
+	// So it can prevent old cache recorded combined with token.
+	disallowCacheWithSameToken bool
```

Comment:
This is to guarantee after a new pushReq, we donot allow cache outdated data from an old version push.